src/theory/datatypes/theory_datatypes_utils.cpp

```diff
@@ -138,26 +139,41 @@ Node mkSygusTerm(const Datatype& dt,
     Assert(children.size() == 1);
     return children[0];
   }
-  if (op.getKind() != BUILTIN)
+  // get the kind of the operator
+  Kind ok = op.getKind();
+  if (ok != BUILTIN)
   {
-    schildren.push_back(op);
+    if (ok==LAMBDA && doBetaReduction)
+    {
+      // do immediate beta reduction
+      std::vector< Node > vars{op[0].begin(),op[0].end()};
+      Assert(vars.size()==children.size());
+      Node ret = op[1].substitute(vars.begin(),vars.end(),children.begin(), children.end());
```

Comment:
shouldn't you be doing the capture avoiding substitution?

Reply:
No, the terms we are using to substitute (those generated by sygus grammars) do not have quantifiers in them, nor does the body of op have quantifiers, so it is safe to use the normal substitution.

Reply:
I see. Could you add a comment over that line to clarify that?

Reply:
Done.